xline/src/storage/kvwatcher.rs

```diff
@@ -0,0 +1,312 @@
+        self.inner.cancel(watcher).await
+    }
+}
+
+impl KvWatcherInner {
+    /// New `KvWatchInner`
+    fn new(storage: Arc<KvStoreInner>) -> Self {
+        Self {
+            storage,
+            watcher_map: Mutex::new(HashMap::new()),
+        }
+    }
+
+    /// Create a watch to KV store
+    async fn watch(
+        &self,
+        id: WatchId,
+        key_range: KeyRange,
+        start_rev: i64,
+        filters: Vec<i32>,
+        event_tx: mpsc::Sender<WatchEvent>,
+    ) -> (Watcher, Vec<Event>, i64) {
+        let watcher = Watcher::new(key_range.clone(), id, start_rev, filters, event_tx);
+        let revision = self.storage.revision();
+        // TODO handle racing that new event is generated before watcher is registered
+        let events = if start_rev == 0 {
+            vec![]
+        } else {
+            self.storage
+                .get_event_from_revision(key_range.clone(), start_rev)
+        };
+
+        let _prev = self
+            .watcher_map
+            .lock()
+            .await
+            .entry(key_range)
+            .and_modify(|set| {
+                let _ = set.insert(watcher.clone());
+            })
+            .or_insert_with(|| {
+                let mut set = HashSet::new();
+                let _ = set.insert(watcher.clone());
+                set
+            });
+        (watcher, events, revision)
+    }
+
+    /// Cancel a watch from KV store
+    async fn cancel(&self, watcher: Watcher) -> i64 {
+        let revision = self.storage.revision();
+        let key_range = watcher.key_range().clone();
+        let _prev = self
+            .watcher_map
+            .lock()
+            .await
+            .entry(key_range)
+            .and_modify(|set| {
+                let _ = set.remove(&watcher);
+            });
+
+        revision
+    }
+
+    /// Handle KV store updates
+    async fn handle_kv_updates(&self, updates: (i64, Vec<Event>)) {
+        let (revision, updates) = updates;
+        let watcher_map = self.watcher_map.lock().await;
+        for (k, v) in watcher_map.iter() {
+            let events: Vec<Event> = updates
+                .clone()
+                .into_iter()
+                .filter(|e| {
+                    k.contains_key(
+                        &e.kv
+                            .as_ref()
+                            .unwrap_or_else(|| panic!("Receive Event with empty kv"))
+                            .key,
+                    )
+                })
+                .collect();
+
+            if !events.is_empty() {
+                for watcher in v.iter() {
+                    if revision < watcher.start_rev() {
+                        continue;
+                    }
+                    let filtered_event = watcher.filter_events(events.clone());
+                    if !filtered_event.is_empty() {
+                        watcher
+                            .notify(watcher.watch_id(), revision, filtered_event)
```

Comment:
Do we have to send the notification while holding the watcher_map lock?

Reply:
We need to generate all data to send during holding the lock, this will cause a lot of copy. And this notify is just sendimg message through async channel.

Reply:
Can we use arc to avoid payload copy?

Reply:
Done